### ISSUE TITLE
streamFile broken since http 3.1.2 behind a load balancer

### ISSUE THREAD

**Issue by vzsg** [vapor/vapor#1801]
Since vapor/http#306 was merged and version was 3.1.2 tagged, `FileMiddleware` and the `streamFile` method (which builds on streamFile as well) are broken when the app is deployed in an environment with a load balancer in front. This includes Heroku and Vapor Cloud.

Regular routes, including Leaf views don't seem to be affected. While the workaround at the end of the issue works, this information suggests that the real problem is with the streamFile method and hopefully it could be fixed to work with the current version of http as well.

### Steps to reproduce

1. Start from the `api-template`.
1. Create a new file as `Public/index.html` with the following contents:
    ```html
    <html>
    <head><title>Hello!</title></head>
    <body>
    <p>Hello World!</p>
    </body>
    </html>
    ```

1. Enable the FileMiddleware in `configure.swift` by uncommenting the relevant line.
1. Add a route handler to `routes.swift` that uses `streamFile`:
    ```swift
    router.get("/") { req -> Future<Response> in
        let dirs: DirectoryConfig = try req.make()
        return try req.streamFile(at: dirs.workDir + "Public/index.html")
    }
    ```

1. Deploy to Heroku or Vapor Cloud.
    ```bash
    echo "web: Run serve --env production --hostname 0.0.0.0 --port $PORT" > Procfile
    git add .
    git commit -m "Initial import"
    heroku apps:create
    heroku buildpacks:set https://github.com/vapor-community/heroku-buildpack.git
    git push heroku master
    heroku ps:scale web=1
    ```

1. Open the deployed site in a browser – this will trigger the / route. If you're using Heroku, you can simply run:
    ```bash
    heroku open
    ```
1. Navigate to `https://theapplications/index.html` to check FileMiddleware.

### Expected behavior

Both the `/` route and `/index.html` should result in "Hello World!" being displayed.

### Actual behavior

Heroku returns an Application Error page for either route:

<img width="566" alt="screen shot 2018-09-08 at 11 03 37" src="https://user-images.githubusercontent.com/1783465/45252459-ebd58b00-b356-11e8-9a34-317dfe0ecda5.png">

And the following lines are printed to the application logs:

```
2018-09-08T08:57:56.981556+00:00 app[web.1]: [ ERROR ] ioOnClosedChannel (/tmp/build_0c0f3c769fa4d2ef6c31398f01187d6c/.build/checkouts/vapor.git-6399936683827946185/Sources/Vapor/Logging/Logger+LogError.swift:32)
2018-09-08T08:57:56.981734+00:00 heroku[router]: at=error code=H13 desc="Connection closed without response" method=GET path="/index.html" host=young-reaches-59886.herokuapp.com request_id=c6a5864a-d2d5-4a9e-94e9-8cbb509984b4 fwd="80.98.65.234" dyno=web.1 connect=0ms service=1ms status=503 bytes=0 protocol=https
```

Vapor Cloud on the other hand returns an empty response with the 502 status code.

### Environment

* Vapor Framework version: 3.0.8 (vapor/http 3.1.2)
* Vapor Toolbox version: N/A
* OS version: Ubuntu 16.04

### Workaround

At this time, the workaround we found is monkey-patching Package.resolved to pull in http 3.1.1 instead of 3.1.2. This is done by editing Package.resolved, searching for the HTTP reference, and updating the `version` and `revision` keys to match:

```json
{
  "package": "HTTP",
  "repositoryURL": "https://github.com/vapor/http.git",
  "state": {
    "branch": null,
    "revision": "0fac5c7c9ab1e424621ba001d83cb59ffca99ad2",
    "version": "3.1.1"
  }
},
```

**Comment by joscdk** [comment]
For info, i just setup a test system that breaks on Cloud 2. On the nginx proxy i get the following output from a simple site that loads an image:

```
[error] 858#858: *917 upstream prematurely closed connection while reading response header from upstream
```

**Comment by vzsg** [comment]
I think this is what Heroku's router is also telling us: the connection was closed before writing a single byte.

**Comment by tanner0101** [comment]
I think I see the problem. Looks like an issue w/ the logic implemented by https://github.com/vapor/http/pull/306 that I didn't catch during review. 

If the request is not keep-alive, the new code closes the connection after the call to serialization. This is fine in all cases _except_ chunked stream where data may continue to be sent _after_ that. `streamFile` is one of the few places where chunked stream is used. 

Working on a fix now.

**Comment by tanner0101** [comment]
Testing potential fix here: https://github.com/vapor/http/pull/307

**Comment by joscdk** [comment]
I just tested stream file fix on nginx in Cloud 2, and it works, all assets are showing :)

**Comment by tanner0101** [comment]
Fixed in vapor/http 3.1.3
